cmd/ocsp-updater/main.go

```diff
@@ -219,42 +333,51 @@ func main() {
 
 		cac, closeChan := setupClients(c, stats)
 
+		updater, err := newUpdater(
+			auditlogger,
+			stats,
+			clock.Default(),
+			dbMap,
+			cac,
+			// Necessary evil for now
+			c.OCSPUpdater,
+		)
+		cmd.FailOnError(err, "Failed to create updater")
+
+		auditlogger.Info(app.VersionString())
+
 		go func() {
-			// Abort if we disconnect from AMQP
 			for {
-				for err := range closeChan {
-					auditlogger.Warning(fmt.Sprintf(" [!] AMQP Channel closed, aborting early: [%s]", err))
-					panic(err)
-				}
+				err := <-closeChan
+				// TODO(): When the channel falls over so do we for now, if the AMQP channel
+				// has already closed there is no real cleanup we can do. This is due to
+				// really needing to change the underlying AMQP Server/Client reconnection
+				// logic.
+				auditlogger.Warning(fmt.Sprintf(" [!] AMQP Channel closed, exiting: [%s]", err))
+				os.Exit(1)
 			}
 		}()
 
-		auditlogger.Info(app.VersionString())
-
-		updater := &OCSPUpdater{
-			cac:   cac,
-			dbMap: dbMap,
-			stats: stats,
-			log:   auditlogger,
-		}
-
-		// Calculate the cut-off timestamp
-		if c.OCSPUpdater.MinTimeToExpiry == "" {
-			panic("Config must specify a MinTimeToExpiry period.")
-		}
-		dur, err := time.ParseDuration(c.OCSPUpdater.MinTimeToExpiry)
-		cmd.FailOnError(err, "Could not parse MinTimeToExpiry from config.")
-
-		oldestLastUpdatedTime := time.Now().Add(-dur)
-		auditlogger.Info(fmt.Sprintf("Searching for OCSP responses older than %s", oldestLastUpdatedTime))
-
-		// When we choose to batch responses, it may be best to restrict count here,
-		// change the transaction to survive the whole findStaleResponses, and to
-		// loop this method call however many times is appropriate.
-		err = updater.findStaleResponses(oldestLastUpdatedTime, c.OCSPUpdater.ResponseLimit)
-		if err != nil {
-			auditlogger.WarningErr(err)
-		}
+		go updater.newCertificatesLoop.loop()
+		go updater.oldOCSPResponsesLoop.loop()
+
+		// Catch INT/TERM/HUP signals and exit after work is done
+		sigChan := make(chan os.Signal, 1)
+		signal.Notify(sigChan, syscall.SIGTERM)
+		signal.Notify(sigChan, syscall.SIGINT)
+		signal.Notify(sigChan, syscall.SIGHUP)
+
+		// Block until signalled
+		<-sigChan
+		signal.Stop(sigChan)
+		auditlogger.Info(" [!] Caught signal, finishing work in progress")
+
+		// We can only acquire a write lock if all read locks have been relinquished so
+		// we wait until then before exiting (this is a bit janky and could probably be
+		// done better with a WaitGroup and closing channels, but it also helps with restarting
+		// so uh... idk)
+		updater.rpcMu.Lock()
```

Comment:
This code is pretty difficult to determine if its used correctly. Can we use launchpad.net/tomb or similar to do explicit checks to make sure we're not stopped while doing work that can't be undone?

But, I am curious to know what work is in this code that we can't just crash the process during because it's probably something we should rework such that we can just crash while its occuring. (Since crashes are likely to just happen)